### PR TITLE
Localisation Wrapper

### PR DESCRIPTION
**Pull request by jojolepro**
Wraps a localisation context in a Asset wrapper.
Slightly inconvenient to use because we have to wait for the asset to be loaded, but I'm not sure how to go around that.

---
This change is [Reviewable](https://reviewable.io/reviews/amethyst/amethyst/663)

### CODE REVIEW

**Comment by jojolepro** [comment]
Just realized I forget to rename things I copy pasted. Don't review now thanks ;)

File: examples/locale/main.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    pub fn new() -> Self {
+        Example {
+            handle: None,
+        }
+    }
+}
+
+impl State for Example {
+    fn on_start(&mut self, world: &mut World) {
+        world.add_resource(AssetStorage::<Locale>::new());
```
Comment (Rhuagh) [inline examples/locale/main.rs]: No longer needed, you add `Processor<Locale>` below, which will register this resource.

**Comment by jojolepro** [comment]
@torkleyy Its not useful in this case as ProgressCounter is made when batch loading multiple assets to prevent checking them one by one.

File: amethyst_assets/src/asset.rs [hunk 2]
```diff
@@ -3,6 +3,20 @@ use std::sync::Arc;
 use amethyst_core::specs::storage::UnprotectedStorage;
 #[cfg(feature = "profiler")]
 #[macro_use]
+#[cfg(feature = "profiler")]
```
Comment (Rhuagh) [inline amethyst_assets/src/asset.rs]: Uuuh, wtf?

Comment (jojolepro) [reply amethyst_assets/src/asset.rs]: can I die?

File: amethyst_assets/src/asset.rs [hunk 3]
```diff
@@ -9,6 +9,14 @@ use amethyst_core::specs::storage::UnprotectedStorage;
 #[macro_use]
 #[cfg(feature = "profiler")]
 #[macro_use]
+#[cfg(feature = "profiler")]
```
Comment (torkleyy) [inline amethyst_assets/src/asset.rs]: I think one of these might be enough ;P

**Comment by Rhuagh** [comment]
Fix the conflicts and fix the review comment above and this is ready for merge.

**Comment by jojolepro** [comment]
#[cfg(feature = "profiler")]
#[macro_use]
#[cfg(feature = "profiler")]
#[macro_use]
#[cfg(feature = "profiler")]
#[macro_use]
#[cfg(feature = "profiler")]
#[macro_use]
#[cfg(feature = "profiler")]
#[macro_use]
#[cfg(feature = "profiler")]
#[macro_use]
#[cfg(feature = "profiler")]
#[macro_use]
#[cfg(feature = "profiler")]
#[macro_use]
use thread_profiler::{register_thread_with_profiler, write_profile};

I'm rolling back the formatting changes.

**Comment by jojolepro** [comment]
Done

File: examples/locale/main.rs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            (),
+            &world.read_resource(),
+        ));
+    }
+
+    fn update(&mut self, world: &mut World) -> Trans {
+        let store = world.read_resource::<AssetStorage<Locale>>();
+
+        // Check if the locale has been loaded.
+        // If you are doing this for multiple assets, you should be using `ProgressCounter`.
```
Comment (torkleyy) [inline examples/locale/main.rs]: This example should show good practices, so please use the counter.

File: examples/locale/main.rs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        if let Some(locale) = self.handle.as_ref().and_then(|h| store.get(h)) {
+            println!(
+                "{}",
+                locale
+                    .context
+                    .get_message("hello")
+                    .and_then(|msg| locale.context.format(msg, None))
+                    .unwrap()
```
Comment (torkleyy) [inline examples/locale/main.rs]: Should not use unwrap

**Comment by jojolepro** [comment]
https://github.com/projectfluent/fluent-rs

**Comment by Moxinilian** [comment]
I've seen that already, and the only way I've seen is to squash all languages in the same file.
Is it the only way? Does it mean we will load the language file asset once and have the system select language accordingly? There is no way to logically separate languages?

**Comment by Rhuagh** [comment]
Conflictomania, but apart from that ready I believe?

File: examples/locale/main.rs (hunk 1 above)
Comment (Rhuagh) [reply examples/locale/main.rs]: Ah, this is not fixed yet.

**Comment by jojolepro** [comment]
Should be

File: examples/locale/main.rs (hunk 1 above)
Comment (torkleyy) [reply examples/locale/main.rs]: This should be using `world.exec`.

**Comment by jojolepro** [comment]
soon :tm:

**Comment by jojolepro** [comment]
Just updating the branch, will try to finish this thing this weekend. I'm not sure what changed last month so I don't know how much changes I need to make yet.

File: Cargo.toml [hunk 6]
```diff
@@ -152,12 +152,17 @@ name = "appendix_a"
 path = "examples/appendix_a/main.rs"
 
 [[example]]
+<<<<<<< HEAD
```
Comment (Xaeroxe) [inline Cargo.toml]: Woops, merge conflict merged

**Comment by jojolepro** [comment]
rustfmt pass 1 sec

**Comment by jojolepro** [comment]
Ready for merge yay. :D
@torkleyy can re-review if he wants to.

**Comment by jojolepro** [comment]
missing changelog entry

**Comment by jojolepro** [comment]
Ready for real now! (I totally forgot I had to update the changelog for a full week :P )

**Pull request merged**